### PR TITLE
serial task and kill if modified

### PR DESCRIPTION
**Pull request by calellowitz**
## Summary

https://dimagi-dev.atlassian.net/browse/SAAS-12510

I am not totally sold on this approach, but I think it may be a step in the right direction and am curious what people think. The PR includes two separate parts.

The first is simply to make this task a serial task with the lock keyed on the data source id so that we will never have two rebuilds running at the same time for the same data source. This is nice both because thats a dangerous situation, having them both modifying the data source, and because it ensures no large data source will monopolize the queue.

The second change is to more directly deal with the issue in the ticket (someone kicking off rebuilds repeatedly that each ran for hours). It makes it so that the automated rebuilds kicked off by the report builder will early exit if the data source has been changed since it was kicked off. This will hopefully mean that when someone kicks off multiple in quick succession (with repeated saving), only the last will actually run.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/userreports/tasks.py
Comment (snopoke) [inline corehq/apps/userreports/tasks.py:95]: won't `trigger_time` always be greater since it was recorded after the config was saved? What about using the `_rev` and only running the task if the rev passed in is equal to the rev of the config? Alternately pass in `config.last_modified` directly instead of using `utcnow()`.

Final thought, perhaps you meant to say `trigger_time < config.last_modified`

**Review by snopoke — COMMENTED** [review]
So this will prevent a new task from starting if the config has been modified since the task was triggered (pending changes from my comment in the PR). 

Do you think it could be worth aborting a running rebuild early if there is a config change? That would allow the latest rebuild to take over instead of having to wait for the previous one to finish first.

**Comment by calellowitz** [comment]
> So this will prevent a new task from starting if the config has been modified since the task was triggered (pending changes from my comment in the PR).

Yep, exactly. So that only the last triggered rebuild, which happens on every save in report builder if there are changes, will end up running.

> Do you think it could be worth aborting a running rebuild early if there is a config change? That would allow the latest rebuild to take over instead of having to wait for the previous one to finish first.

I looked at that. I was a little hesitant about having it refetch the doc frequently during the process, and I think we would still want to restrict it to report builder rebuilds, but it doesn't seem like a big change if you prefer it that way. One other consideration is that stopping mid rebuild could put that data source in a weird state for an unknown period of time. While a rebuild is happening the data is incomplete, but the user can at least see it filling in and so knows its happening. If that suddenly stops, but the next rebuild is delayed due to other items in the queue, that could look confusing, and perhaps encourage the user the try to fire off more rebuilds.

**Review by dannyroberts — COMMENTED** [review]
The approach seems good to me. The effect would seem to be that only one rebuild task can run at a time for a given report, and only the latest available task will ever start, but once it does it will run to completion. That seems as good a heuristic as we're going to get in this first pass for balancing the resources that we dedicate to a single user's activity versus the responsiveness that that user gets (while also avoiding undefined behavior).

Had just one comment, but I think after that I'm ready to approve

**Review by snopoke — APPROVED** [review]

File: corehq/apps/userreports/tasks.py
Comment (dannyroberts) [reply corehq/apps/userreports/tasks.py:92]: Sounds reasonable. I'll leave it up to you whether you want to change the timeout and/or delay settings

**Review by dannyroberts — APPROVED** [review]

**Pull request merged**
